spec/const3.dd

```diff
@@ -262,14 +262,18 @@ char[] p = "world"; // error, cannot implicitly convert immutable
 
         $(P
         The second way is to cast data to immutable.
-        When doing so, it is up to the programmer to ensure that no
-        other mutable references to the same data exist.
+        When doing so, it is up to the programmer to ensure that any mutable
+        references to the same data are not used to modify the data after the
+        cast.
```

Comment:
This is meant as a relaxation, right? Data cast to immutable can still be read legally through mutable references, and as long as that's done things should still be fine. I think this works (famous last words...). cc @tgehr @WalterBright

Reply:
Yes, but maybe it should be clarified what "ensure" means. In particular, those mutable references should still not even be accessible from @safe code. (Even if the current version of the @safe code does not use them to write.)

Reply:
> maybe it should be clarified what "ensure" means. In particular, those mutable references should still not even be accessible from @safe code. (Even if the current version of the @safe code does not use them to write.)

I tried covering that in the definition of `@trusted` ("cannot lead to undefined behavior in @safe code that is executed afterwards"). I'm not sure if I managed to express what's in your parentheses, though. I've got it in an example, but in the prose it might not be clear enough.

Reply:
The point of @safe code is to offer guarantees, and allowing @safe code to continue to have mutable references to immutable data means that the compiler has no way to guarantee the @safe code isn't attempting mutation. I don't see a purpose to this relaxation, either, i.e. what important use case would it enable?

Reply:
> The point of @safe code is to offer guarantees, and allowing @safe code to continue to have mutable references to immutable data means that the compiler has no way to guarantee the @safe code isn't attempting mutation.

Allowing that is not my goal. I'm trying to:

1. allow creating such references in `@system`/`@trusted` code (not `@safe` code), and
2. disallow leaking dangerous stuff like that from `@trusted` code to `@safe` code.

Point 1 is a relaxation of the rules, done to fix an inconsistency in the spec ([issue 19621](https://issues.dlang.org/show_bug.cgi?id=19621)).
Point 2 is a tightening of the rules, done to fix a safety hole. Note that the hole is not being opened by point 1; it's there regardless.

If it makes for easier review, I could split this PR in two, as the two points are mostly independent of each other. Let me know if that would help.

> I don't see a purpose to this relaxation, either, i.e. what important use case would it enable?

The spec currently allows [casting `immutable` away](https://dlang.org/spec/const3.html#removing_with_cast). The motivating example given for that is calling a function where a parameter is de-facto `const`, but it's not annotated as such.

Similarly, you might want to call a function where a parameter is de-facto `const`, but it's marked as `immutable`. As far I can see, GtkD, a popular library, is full of such functions. Here's one example: [Window.setTitle](https://github.com/gtkd-developers/GtkD/blob/1349b9e8bd78f3bbb50b8cb1fddc48d8e4414226/generated/gtkd/gtk/Window.d#L2300-L2303)